### PR TITLE
Fetch whether task is interruptible as part of taskMetaData

### PR DESCRIPTION
**Pull request by migueltol22**
# TL;DR
Add IsInterruptible to TaskExecutionMetadata Interface. This will help us identify pods that are OK being evicted so we can place them on cheaper machines but higher chance of getting terminated.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Based on the value of this flag, plugins add appropriate labels and tolerations to the k8s resources so that pods are placed on right set of machines (eg: pod with interruptible=true can go to a spot instance when cluster is running low on reserved capacity)

## Tracking Issue
https://github.com/lyft/flyte/issues/207

## Follow-up issue
_NA_

### CODE REVIEW

File: go/tasks/pluginmachinery/core/exec_metadata.go [hunk 1]
```diff
@@ -32,4 +32,5 @@ type TaskExecutionMetadata interface {
 	GetLabels() map[string]string
 	GetAnnotations() map[string]string
 	GetK8sServiceAccount() string
+	GetInterruptible() bool
```
Comment (surindersinghp) [inline go/tasks/pluginmachinery/core/exec_metadata.go]: I like "IsInterruptible" ("Is...") for boolean properties.

**Comment by surindersinghp** [comment]
cc @kumare3 @EngHabu

File: go/tasks/pluginmachinery/core/exec_metadata.go (hunk 1 above)
Comment (migueltol22) [reply go/tasks/pluginmachinery/core/exec_metadata.go]: Updated.

Comment (surindersinghp) [reply go/tasks/pluginmachinery/core/exec_metadata.go]: Thank you.

File: go/tasks/pluginmachinery/flytek8s/k8s_resource_adds.go [hunk 2]
```diff
@@ -124,6 +124,34 @@ func DecorateEnvVars(ctx context.Context, envVars []v1.EnvVar, id pluginsCore.Ta
...
+	// 2. Get the tolerations for interruptible pods
+	if interruptible && config.GetK8sPluginConfig().InterruptibleTolerations != nil{
+		tolerations = append(tolerations, config.GetK8sPluginConfig().InterruptibleTolerations...)
+	}
+
+	return tolerations
+}
+
+
 func GetTolerationsForResources(resourceRequirements ...v1.ResourceRequirements) []v1.Toleration {
```
Comment (akhurana001) [inline go/tasks/pluginmachinery/flytek8s/k8s_resource_adds.go]: can we merge with the above newly created method and avoid logic duplication ?

Comment (surindersinghp) [reply go/tasks/pluginmachinery/flytek8s/k8s_resource_adds.go]: it was already merged :), this came as a bad merge. will clean it.

Comment (surindersinghp) [reply go/tasks/pluginmachinery/flytek8s/k8s_resource_adds.go]: this is still WIP. I will push some code change + tests in a bit.

**Comment by kumare3** [comment]
@migueltol22 @surindersinghp can we please fill out the description correctly. Please add 
- Complete description
and create a
- Tracking Issue

cc @EngHabu please do not +1 unless we get these things added. We will have to make it a habit.

File: go/tasks/pluginmachinery/flytek8s/config/config.go [hunk 3]
```diff
@@ -43,6 +43,10 @@ type K8sPluginConfig struct {
 	DefaultCPURequest string `json:"default-cpus" pflag:",Defines a default value for cpu for containers if not specified."`
 	// default memory requests for a container
 	DefaultMemoryRequest string `json:"default-memory" pflag:",Defines a default value for memory for containers if not specified."`
+	// Tolerations for interruptible pods
```
Comment (kumare3) [inline go/tasks/pluginmachinery/flytek8s/config/config.go]: can we explain what interruptible means?

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go [hunk 4]
```diff
@@ -202,7 +216,10 @@ func ConvertPodFailureToError(status v1.PodStatus) (string, string) {
 		if containerState.Terminated != nil {
 			if strings.Contains(c.State.Terminated.Reason, OOMKilled) {
```
Comment (kumare3) [inline go/tasks/pluginmachinery/flytek8s/pod_helper.go]: lets add some documentation here, why we are doing this

File: go/tasks/pluginmachinery/flytek8s/config/config.go [hunk 5]
```diff
@@ -43,6 +43,10 @@ type K8sPluginConfig struct {
...
+	InterruptibleTolerations []v1.Toleration `json:"interruptible-tolerations"  pflag:"-,Tolerations to be applied for interruptible pods"`
+	// Node Labels for interruptible pods
+	InterruptibleLabels map[string]string `json:"interruptible-labels" pflag:"-,Defines a set of labels to add to the interruptible pods."`
```
Comment (EngHabu) [inline go/tasks/pluginmachinery/flytek8s/config/config.go]: Should this be a `InterruptibleNodeSelector` instead? I don't think it has to be labels...

File: go/tasks/pluginmachinery/flytek8s/pod_helper_test.go [hunk 6]
```diff
@@ -47,7 +50,7 @@ func dummyTaskExecutionMetadata(resources *v1.ResourceRequirements) pluginsCore.
 	to := &pluginsCoreMock.TaskOverrides{}
 	to.On("GetResources").Return(resources)
 	taskExecutionMetadata.On("GetOverrides").Return(to)
-
+	taskExecutionMetadata.On("IsInterruptible").Return(true)
```
[the remark this answers is not in the document]
Comment (surindersinghp) [reply go/tasks/pluginmachinery/flytek8s/pod_helper_test.go]: I am keeping the same style as rest of the methods (for better readability). Wherever we have the new format, I am using this style.

File: go/tasks/pluginmachinery/flytek8s/config/config.go (hunk 3 above)
Comment (surindersinghp) [reply go/tasks/pluginmachinery/flytek8s/config/config.go]: sure

File: go/tasks/pluginmachinery/flytek8s/pod_helper.go (hunk 4 above)
Comment (surindersinghp) [reply go/tasks/pluginmachinery/flytek8s/pod_helper.go]: sure

File: go/tasks/plugins/k8s/sidecar/sidecar.go [hunk 7]
```diff
@@ -60,7 +60,7 @@ func validateAndFinalizeContainers(
 
 	}
 	pod.Spec.Containers = finalizedContainers
-	pod.Spec.Tolerations = flytek8s.GetTolerationsForResources(resReqs...)
+	pod.Spec.Tolerations = flytek8s.GetPodTolerations(taskCtx.TaskExecutionMetadata().IsInterruptible(), resReqs...)
```
[the remark this answers is not in the document]
Comment (surindersinghp) [reply go/tasks/plugins/k8s/sidecar/sidecar.go]: done

File: go/tasks/plugins/hive/test_helpers.go [hunk 8]
```diff
@@ -68,6 +68,7 @@ func GetMockTaskExecutionMetadata() core.TaskExecutionMetadata {
 		Kind: "node",
 		Name: "blah",
 	})
+	taskMetadata.On("IsInterruptible").Return(true)
```
[the remark this answers is not in the document]
Comment (surindersinghp) [reply go/tasks/plugins/hive/test_helpers.go]: Hey, I am trying to keep the same style as rest of this method. We can update all the tests later. 
@EngHabu is that OK?

[the remark this answers is not in the document]
Comment (EngHabu) [reply go/tasks/plugins/hive/test_helpers.go]: We never will :D you know that, right? but that's up to you...

**Comment by surindersinghp** [comment]
@EngHabu @kumare3 do you guys mind taking another look?
